cpp/src/io/parquet/compact_protocol_writer.cpp

```diff
@@ -0,0 +1,247 @@
+  }
+  return c.value();
+}
+
+size_t CompactProtocolWriter::write(const SchemaElement &s)
+{
+  CompactProtocolFieldWriter c(*this);
+  if (s.type != UNDEFINED_TYPE) {
+    c.field_int(1, s.type);
+    if (s.type_length != 0) { c.field_int(2, s.type_length); }
+  }
+  if (s.repetition_type != NO_REPETITION_TYPE) { c.field_int(3, s.repetition_type); }
+  c.field_string(4, s.name);
+
+  if (s.type == UNDEFINED_TYPE) { c.field_int(5, s.num_children); }
+  if (s.converted_type != UNKNOWN) {
+    c.field_int(6, s.converted_type);
+    if (s.converted_type == DECIMAL) {
+      c.field_int(7, s.decimal_scale);
+      c.field_int(8, s.decimal_precision);
+    }
+  }
+  return c.value();
+}
+
+size_t CompactProtocolWriter::write(const RowGroup &r)
+{
+  CompactProtocolFieldWriter c(*this);
+  c.field_struct_list(1, r.columns);
+  c.field_int(2, r.total_byte_size);
+  c.field_int(3, r.num_rows);
+  return c.value();
+}
+
+size_t CompactProtocolWriter::write(const KeyValue &k)
+{
+  CompactProtocolFieldWriter c(*this);
+  c.field_string(1, k.key);
+  if (k.value.size() != 0) { c.field_string(2, k.value); }
+  return c.value();
+}
+
+size_t CompactProtocolWriter::write(const ColumnChunk &s)
+{
+  CompactProtocolFieldWriter c(*this);
+  if (s.file_path.size() != 0) { c.field_string(1, s.file_path); }
+  c.field_int(2, s.file_offset);
+  c.field_struct(3, s.meta_data);
+  if (s.offset_index_length != 0) {
+    c.field_int(4, s.offset_index_offset);
+    c.field_int(5, s.offset_index_length);
+  }
+  if (s.column_index_length != 0) {
+    c.field_int(6, s.column_index_offset);
+    c.field_int(7, s.column_index_length);
+  }
+  return c.value();
+}
+
+size_t CompactProtocolWriter::write(const ColumnChunkMetaData &s)
+{
+  CompactProtocolFieldWriter c(*this);
+  c.field_int(1, s.type);
+  c.field_int_list(2, s.encodings);
+  c.field_string_list(3, s.path_in_schema);
+  c.field_int(4, s.codec);
+  c.field_int(5, s.num_values);
+  c.field_int(6, s.total_uncompressed_size);
+  c.field_int(7, s.total_compressed_size);
+  c.field_int(9, s.data_page_offset);
+  if (s.index_page_offset != 0) { c.field_int(10, s.index_page_offset); }
+  if (s.dictionary_page_offset != 0) { c.field_int(11, s.dictionary_page_offset); }
+  if (s.statistics_blob.size() != 0) { c.field_struct_blob(12, s.statistics_blob); }
+  return c.value();
+}
+
+void CompactProtocolFieldWriter::putb(uint8_t v) { writer.m_buf.push_back(v); }
+
+void CompactProtocolFieldWriter::putb(const uint8_t *raw, uint32_t len)
+{
+  for (uint32_t i = 0; i < len; i++) writer.m_buf.push_back(raw[i]);
+}
+
+uint32_t CompactProtocolFieldWriter::put_uint(uint64_t v)
+{
+  int l = 1;
+  while (v > 0x7f) {
+    putb(static_cast<uint8_t>(v | 0x80));
+    v >>= 7;
+    l++;
+  }
+  putb(static_cast<uint8_t>(v));
+  return l;
+}
+
+uint32_t CompactProtocolFieldWriter::put_int(int64_t v)
+{
+  int64_t s = (v < 0);
+  return put_uint(((v ^ -s) << 1) + s);
+}
+
+void CompactProtocolFieldWriter::put_fldh(int f, int cur, int t)
```

Comment:
I'd request some documentation for these. Or at least more descriptive names. This one seems to be for writing field header?